worker/dbaccessor/worker.go

```diff
@@ -9,16 +9,31 @@ import (
 	"os"
 	"path/filepath"
 	"sync"
+	"time"
 
 	"github.com/juju/clock"
 	"github.com/juju/errors"
+	"github.com/juju/retry"
 	"github.com/juju/worker/v3"
 	"github.com/juju/worker/v3/catacomb"
+	"gopkg.in/tomb.v2"
 
+	coredb "github.com/juju/juju/core/db"
 	"github.com/juju/juju/database/app"
 )
 
-const replSocketFileName = "juju.sock"
+const (
+	// PollInterval is the amount of time to wait between polling the database.
+	PollInterval = time.Second * 30
+)
+
+// TrackedDB defines the union of a TrackedDB and a worker.Worker interface.
+// This is local to the package, allowing for better testing of the underlying
+// trackerDB worker.
+type TrackedDB interface {
+	coredb.TrackedDB
```

Comment:
Should we make this a func that closes over pulling the DB instance out of the cache?
Pro: late acquisition means robust behaviour if the worker decides to nuke a DB instance and reacquire it.
Con: we can't really rely on long lived prepared statements.

Reply:
> we can't really rely on long lived prepared statements.

You can't anyway, as you can't guarantee that the db you have is the same one you've been given.

Reply:
Actually, I have found a way to do this. Essentially, we register a prepare statement hook, which is called initially, then every time a new DB comes online. That way we can guarantee that a statement is correctly associated with the right db.

I don't think this will be used for a lot of cases, probably just for hot paths.

```go
var stmts map[string]*sql.Stmt
_, _ = trackedDB.PrepareStmts(func(db *sql.DB) error {
	stmts = make(map[string]*sql.Stmt)

	stmt, err := db.Prepare("SELECT * FROM leases;")
	if err != nil {
		return errors.Trace(err)
	}
	stmts["lease"] = stmt
	return nil
})

trackedDB.Txn(context.TODO(), func(ctx context.Context, tx *sql.Tx) error {
	stmt, ok := stmts["lease"]
	if !ok {
		return errors.Errorf("no stmt")
	}

	stmt = tx.Stmt(stmt)
	rows, err := stmt.Query()
	...
})
```